File: go/src/RumbleRacing/core/asset/chunk.go
package asset

import (
	"encoding/binary"
	"fmt"
	"rumble-reader/helpers"
)

type Chunk struct {
	Offset  int
	Magic   [4]byte
	Size    uint32
	Payload []byte
}

func (c Chunk) MagicString() string {
	return string(c.Magic[:])
}

func ParseChunks(data []byte) ([]Chunk, error) {
	var chunks []Chunk
	offset := 0
	for offset < len(data) {
		if offset+8 > len(data) {
			return nil, fmt.Errorf("incomplete chunk header at offset %d", offset)
		}
		var magic [4]byte
		copy(magic[:], data[offset:offset+4])
		size := binary.LittleEndian.Uint32(data[offset+4 : offset+8])
		if size < 8 {
			return nil, fmt.Errorf("invalid chunk size %d for %q", size, magic)
		}
		chunkEnd := offset + int(size)
		if chunkEnd > len(data) {
			return nil, fmt.Errorf("chunk %q size %d exceeds remaining data", magic, size)
		}
		payload := data[offset:chunkEnd]

		helpers.ReverseBytesInPlace(magic[:])

		chunks = append(chunks, Chunk{
			Magic:   magic,
			Size:    size,
			Payload: payload,
		})
		offset = chunkEnd
	}
	return chunks, nil
}

File: src/RumbleRacing/asset/chunk.ts
import ArrayBufferSlice from "../../ArrayBufferSlice";
import { readFourCC } from "../helpers/fourCC";

export interface AssetChunk {
  offset: number;
  magic: ArrayBufferSlice;
  size: number;
  payload: ArrayBufferSlice;
}

export function parseChunks(data: ArrayBufferSlice): AssetChunk[] {
  const view = data.createDataView();
  const chunks: AssetChunk[] = [];
  let offset = 0;

  while (offset < data.byteLength) {
    if (offset + 8 > data.byteLength) {
      throw new Error(`incomplete chunk header at offset ${offset}`);
    }
    const magic = data.subarray(offset, 4);
    const size = view.getUint32(offset + 4, true);
    if (size < 8) {
      const tag = readFourCC(magic, 0);
      throw new Error(`invalid chunk size ${size} for "${tag}"`);
    }
    const chunkEnd = offset + size;
    if (chunkEnd > data.byteLength) {
      throw new Error(`chunk size ${size} exceeds remaining data`);
    }
    const payload = data.subarray(offset, size);

    chunks.push({ offset, magic, size, payload });
    offset = chunkEnd;
  }

  return chunks;
}
